nitro-protocol/contracts/auxiliary/BatchOperator.sol

```diff
@@ -0,0 +1,64 @@
+pragma solidity 0.8.17;
+import '../MultiAssetHolder.sol';
+
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+string constant lengthsErr = 'Array lengths must match';
+
+/**
+@dev This contract is used to batch deposit ERC20 tokens into .
+ */
+contract BatchOperator {
+    using SafeERC20 for IERC20;
+    MultiAssetHolder public adjudicator;
+
+    constructor(address adjudicatorAddr) {
+        adjudicator = MultiAssetHolder(adjudicatorAddr);
+    }
+
+    /**
+     * @dev Deposits ETH (or native token for other runtime) into the adjudicator for multiple channels.
+     */
+    function deposit_batch_eth(
+        bytes32[] calldata channelIds,
+        uint256[] calldata expectedHelds,
+        uint256[] calldata amounts
+    ) external payable virtual {
+        require(
+            channelIds.length == expectedHelds.length && expectedHelds.length == amounts.length,
+            lengthsErr
+        );
+        for (uint256 i = 0; i < channelIds.length; i++) {
+            adjudicator.deposit{value: amounts[i]}(
+                address(0),
+                channelIds[i],
+                expectedHelds[i],
+                amounts[i]
+            );
+        }
+    }
+
+    /**
+     * @dev Deposits ERC20 tokens into the adjudicator for multiple channels.
+     */
+    function deposit_batch_erc(
+        address asset,
+        bytes32[] calldata channelIds,
+        uint256[] calldata expectedHelds,
+        uint256[] calldata amounts,
+        uint256 totalAmount
+    ) external payable virtual {
+        require(
+            channelIds.length == expectedHelds.length && expectedHelds.length == amounts.length,
+            lengthsErr
+        );
+
+        IERC20(asset).safeTransferFrom(msg.sender, address(this), totalAmount);
+        IERC20(asset).safeApprove(address(adjudicator), totalAmount);
```

Comment:
I'm wondering if there is a cheaper (but still safe) way of doing this. What if the adjudicator could pull tokens from a supplied address, rather than `msg.sender`? Then, the tokens wouldn't need to go to the batch depositor before going to the adjudicator?

Reply:
Tempting to consider `tx.origin` but that is not recommended https://consensys.github.io/smart-contract-best-practices/development-recommendations/solidity-specific/tx-origin/

Reply:
I was somewhat unsatisfied with this, but in the end the cost is amortized against the entirety of the batch so even at n=2 the total number of approve / transfer operations reaches parity against existing deposit flow.

Your first option is neat, but requires an addition to `deposit`'s fields list which in all other cases would be wasted space.

In general, through various drafts I found that there are a large number of public facing api methods one *could* write in the name of optimizing costs of particular deposit workflows. EG, why not have a `depositETH` function with no asset qualifier?

In the end, I think I'm happy enough to take reasonable gains via methods that don't alter the existing API.

Reply:
That's fair!